src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -3531,6 +3531,42 @@ private uint getExpectedTargetArchitecture()
             }
         }
 
+        private bool doesFieldBelongToClass(CORINFO_FIELD_STRUCT_* fld, CORINFO_CLASS_STRUCT_* cls)
+        {
+            var field = HandleToObject(fld);
+            var queryType = HandleToObject(cls);
+
+            Debug.Assert(!field.IsStatic);
+
+            // doesFieldBelongToClass implements the predicate of...
+            // if field is not associated with the class in any way, return false.
+            // if field is the only FieldDesc that the JIT might see for a given class handle
+            // and logical field pair then return true. This is needed as the field handle here
+            // is used as a key into a hashtable mapping writes to fields to value numbers.
+            //
+            // In this implmentation this is made more complex as the JIT is exposed to CORINFO_FIELD_STRUCT
```

Comment:
Maybe make it clearer that the crossgen2 behavior is different than the runtime behavior?

Is this something we should fix down the road? Seems like it would make us more conservative in places when doing R2R codegen.

Reply:
>> Maybe make it clearer that the crossgen2 behavior is different than the runtime behavior?

I will add a comment.

>> Is this something we should fix down the road? Seems like it would make us more conservative in places when doing R2R codegen.

I hope we will target to get rid of FldSeq and its dependency on FIELD_HANDLE and STRUCT_HANDLE.

If not we can do:
```
if (crossgen2)
{
 canonFieldHandle = JitEEInterface::getCanonField(fieldHandle);
  bool doesBelong = JitEEInterface::doesFieldBelongToClass(canonFieldHandle , clsHnd);
}
```

but I don't like such a solution, it hammers JIT to VM behavior, we can do it always, not only for crossgen2, but then it is an unnecessary VM call.